Update `->Timestamp` to include `report-timezone`

senior opened the pull request:
Specifically this change fixes an issue where `->Timestamp` was
parsing a date string without the `report-timezone` and thus
interpretting the date string in the JVM's timezone instead of the
report timezone.

Although this fixes an important issue, there are several things I don't like about the solution. These conversions are somewhat nuanced and surprisingly complicated. It would be great if there was a way to encapsulate this nuance and make the correct decision automatic.

## Conversion from
Depending on what we're converting from we might or might not need an associated timezone. As an example, if we're converting from an integer/long we don't. Integer/long values are assumed to be in UTC. If we're converting from a `java.util.Date` the date will already be in the JVM's timezone, so no need to have a timezone there either. If we have a String we probably do need a timezone. JodaTime will parse the date as UTC unless the string includes the timezone. This difference makes the much simpler `->Timestamp` function signature more complicated. To facilitate this, I've made a 1-arity and 2-arity version of the function, with a `:pre` condition that will fail if trying to convert a String without a timezone. Not great, but does enforce the constraint.

## Where do I get the timezone
Unfortunately this is somewhat nuanced and we solve this in different ways in different areas of the app. The rules are, if we are parsing a datetime that is part of a query, the database we're querying against supports a `report-timezone` and the user has configured one, we should use that. That `report-timezone` could be coming from the query map (one of our middleware functions add it) or it might be coming from call to the settings table. Sometimes it's bound to a dynamic var, sometimes used directly. If we have no report timezone (or the database doesn't support it) we should use the JVM timezone. Typically this comes from the system property `user.timezone`. Finally the third way is some databases are always in UTC (i.e. BigQuery), for that we always just include the UTC timezone.

## Converting to
I'm not entirely sure that we should be using a Timestamp everywhere that we're using it. We aren't really using the features of the Timestamp (support for nanos) but rather we are basically leaning on the fact that Timestamp is a subclass of Date. I'm not aware of any specific problems that this is causing us, however the [docs say not to do that](https://docs.oracle.com/javase/7/docs/api/java/sql/Timestamp.html) :man_shrugging:.

salsakran commented on src/metabase/query_processor/middleware/cache_backend/db.clj:
```diff
@@ -14,15 +14,17 @@
     :updated_at [:<= (u/->Timestamp (- (System/currentTimeMillis)
-                                       (* 1000 (public-settings/query-caching-max-ttl))))]))
+                                       (* 1000 (public-settings/query-caching-max-ttl)))
+                                    "UTC")]))
```
Does this assume that the server is always running in UTC?

senior replied:
The UTC here is redundant. Since it's using `System/currentTimeMillis`, that function returns a millis since epoch in UTC. Since it's a long, timezone doesn't matter in that `->Timestamp` function. This query will be against the application database and I haven't spent much time looking into our application database and the queries against it. I think it should handle the parameters correctly, but would be good to confirm with a test.

salsakran commented on src/metabase/query_processor/middleware/resolve.clj:
```diff
@@ -230,11 +230,8 @@
-          parsed-string-date (some-> value
-                                     (u/str->date-time tz)
-                                     u/->Timestamp)]
+    (let [tz (or (setting/get :report-timezone) (System/getProperty "user.timezone"))
```
This feels repeated from  https://github.com/metabase/metabase/pull/7520/files#diff-90aba221a9b8d7ad86ce77cad0535f91R601 and seems  like it should be abstracted out somewhere.

salsakran commented on src/metabase/util.clj:
```diff
@@ -44,28 +45,64 @@
-  Date       (->Timestamp [this]
-               (Timestamp. (.getTime this)))
+  nil
+  (coerce-to-timestamp [_]
```
style aside: this is harder to read than the previous formatting.

camsaul replied:
+1 I am not confused by `(->Timestamp something)`

salsakran commented on src/metabase/util.clj:
```diff
@@ -174,7 +211,7 @@
   (boolean (when (string? s)
              (ignore-exceptions
-               (->Timestamp s)))))
+               (->Timestamp s utc)))))
```
why are we defaulting to utc here?

senior replied:
The function returns `true` if the string is parseable as a date, so the results of the parse are discarded. `->Timestamp` needs a timezone to parse a string, so I just hard coded it to UTC

camsaul commented on src/metabase/driver/bigquery.clj:
```diff
@@ -147,21 +147,16 @@
-   (u/->Timestamp s)
+   (u/->Timestamp s time/utc)
    ;; If parsing as ISO-8601 fails parse as a double then convert to ms. Add the appropriate number of milliseconds to
    ;; the number to convert it to the local timezone. We do this because the dates come back in UTC but we want the
```
Don't forget to update the comments explaining what's going on! :)

camsaul commented on src/metabase/util.clj:
```diff
@@ -44,28 +45,64 @@
+  (coerce-to-timestamp [this]
+    (coerce/to-timestamp this)))
+
+(defn ->Timestamp
```
Oh I see 👀 

So does `coerce-to-timestamp` just power `->Timestamp`? Not intended for use elsewhere? We should make it `^:private` if so

salsakran commented:
general feedback: this feels a little too patchwork-y for me to sleep well. I’d like to step up a level or two of abstraction:
1. All times that go through our system have a timezone attached. I really like that this closes places where we were implicitly using jvm’s timezone rather than explicitly using a timezone the user provides
2. IMO we need to distinguish clearly between timezone of the data, and the desired timezone of display. If no one objects `Result Timezone` could work for the latter and `Data Timezone` for the former? I also think this should be reflected in the code itself, as I think overloading timezone to cover both is getting us in trouble overall
3. All time data has a `data timezone`, explicit or implicit. All transformations should be done on a time data object that has a timezone. This should include dates and times stored as longs/etc. 
3a. Sometimes this timezone is implicit, and we provide guesses (but we should also let the user set this)
4. All results from a query process should be mapped to the `result timezone`. 
4a. We can default to something reasonable (like server timezone), but should respect the users choices
4b. All time extracts should operate in the `Result Timezone`

camsaul commented on test/metabase/api/activity_test.clj:
```diff
@@ -22,19 +22,19 @@
                                            :model     "user"
                                            :details   {}
-                                           :timestamp (u/->Timestamp "2015-09-10T05:33:43.641Z")}]]
+                                           :timestamp (u/->Timestamp #inst "2015-09-10T05:33:43.641Z")}]]
```
so `->Timestamp` doesn't work on strings anymore? The idea of that function was to take anything that could conceivably represent a timestamp and turn it into a `Timestamp` object... ISO-8601 Strings represent datetimes...

salsakran commented:
`If we have no report timezone (or the database doesn't support it) we should use the JVM timezone.`

As a further point of terminology, can we agree to always have a `report-timezone`?  It should have a sane default (eg, the jvm timezone), but it should be considered a setting that is set on install and stays consistent. The reason I say install and not at query run time is that if I have a metabase instance and I somehow spin up the jar on a different env with a different timezone, my charts shouldn't all change.

salsakran commented:
As an addendum to the last comment, to keep things sane in case we ever have per-user or per-query timezones, we should always thread a timezone along with each query. 

The QP pipeline would always expect a `report-time` zone. It should never do any kind of default logic to guess it.

iethree committed:
Add a new report-timezone dynamic var, use it as a default [ci drivers]

This commit adds a new `*report-timezone*` dynamic var that should get
bound before dates get parsed (i.e. in a query pipeline
middleware). This allows an easier API `(->Timestamp my-date-string)`
while still ensuring the dates are interpretted correctly.

This is a step toward centralization of our timezone and date
conversion logic. As of this commit there are still separate/similar
logic for handling reporting timezone etc in Pulses, drivers etc.

Fixes #7584, fixes #6402

salsakran commented on src/metabase/driver/bigquery.clj:
```diff
@@ -148,21 +148,13 @@
-                      (.getDSTSavings default-timezone)
-                      (.getRawOffset  default-timezone)))))
+   (du/->Timestamp s time/utc)
+   ;; If parsing as ISO-8601 fails parse as a double then convert to ms. This is ms since epoch in UTC. By using
```
👍

salsakran commented on src/metabase/query_processor.clj:
```diff
@@ -15,6 +15,7 @@
              [add-settings :as add-settings]
              [annotate-and-sort :as annotate-and-sort]
              [binning :as binning]
+             [bind-effective-timezone :as bind-timezone]
```
can we comment what the effective timezone entails? Is it the report timezone, the data timezone, the timezone my grandmother is currently in (helpfully fetched from `whereisgrandma.com/timezone`) so I can be sure not to call her too late?

senior replied:
That might not be the right term for this, but I was trying to come up with a way to refer to to the function/macro that setup "all the timezones". It binds the `report-timezone` if available, or JVM timezone if not. It also binds the data timezone. I can put comments around it, but if there's a better or more descriptive name I'd be :+1: on that too.

salsakran commented on src/metabase/util/date.clj:
```diff
@@ -13,32 +13,113 @@
+(def ^:private jvm-timezone
+  (delay (coerce-to-timezone (System/getProperty "user.timezone"))))
+
+(defn- warn-on-timezone-conflict
```
Like this a lot, we should really figure out a way to warn admins in-app. 

paging @mazameli @kdoh

iethree committed:
Put together a report and database timezone test suite [ci drivers]

The goal is to make it easier to write tests with different data and
report timezone scenarios easier. The tests already have found issues
with our current code and their handling of these timezone
conversions. These tests exposed a bug in MySQL date support which is
also fixed in this commit.

Fixes #6554, fixes #6994, fixes #7393

camsaul reviewed (approved):
This looks pretty solid to me

senior commented:
@camsaul This is ready for another round of review. The approach (centralized, dynamic var etc) has not been changed since you reviewed it earlier. What I have added is the fix for the MySQL native query bug, which fixes several GH issues and added a focused set of timezone related tests that exposed these issues in the simplest way I could think of.

The focused timezone tests are intended to be simpler that the date bucketing test. It doesn't use/convert from unix timestamps, and it's primarily related to filtering using dates and checking the query results have the correct time (in the correct timezone). Data is always loaded in UTC, so there's no ambiguity in how the data is stored and what the correct results should be (if there are no bugs). I think what I have here is useful start of these tests, but the tests are just for BigQuery, PostgreSQL and MySQL right now as those have been causing the most user issue. BigQuery also hits "doesn't support a report timezone" driver case that has been the source of issues as well. I think we'll add to this over time, probably Redshift being next. Adding more backends is sometimes easy, sometimes difficult and I thought I shouldn't hold up fixing these issues on adding more.

iethree committed:
Force connection pool reset before running db tz tests [ci drivers]

It's possible that database connections checked back into the pool
could still have session timezones associated with them. This can
cause the database timezone tests to failure. This will reset the
connection pool before the test runs to ensure that doesn't happen.

camsaul reviewed (approved):
WFM

gunjansriv86 commented:
Today is 2nd of November and i am working in IST timezone.

I found above issue to be in Native query questions. Following is what i have faced while deploying a client's report

1.  **Custom query Builder**. - when we do a filter on date field here  and select day as "Today", it perfectly displays all transactions of 2nd November
![custom query-date filter is correct](https://user-images.githubusercontent.com/44700044/47912413-96cd7580-debe-11e8-96e1-dbca6721b95a.jpg)

2. **Native query builder** - I wrote an SQL and applied dynamic filters on date. when i select day as "Today", it displays all transactions of 1st November.

![native query question-date filter off 1 day](https://user-images.githubusercontent.com/44700044/47912421-a3ea6480-debe-11e8-9ee9-121276fa7660.jpg)


Not sure what is causing this issue? If timezone was the issue, shouldn't it have occured in both the approach

Any suggestions to fix this issue?

Thanks
Gunjan

Matthew-Byrne commented:
Any fixes on the above we are having the same issue but on V.0.33 connected to an Amazon Redshift warehouse and running Postgresql? When I select all the dates in June I have my minimum date as 31st May and the max date as 29th June?
